# Contents List

## 01. Importing Libraries
## 02. Paths
## 03. Importing Data
## 04. Data Manipulation
## 05. Grouping Data with pandas
## 06. Aggregating Data
## 07. Aggregating Data with transform()
## 08. Deriving Columns with loc()
## 09. Task 4.8

# 01. Importing libraries

In [2]:
#Import libraries
import pandas as pd
import numpy as np
import os

# 2. Paths

In [4]:
#Creating a path varaible - Shortcut for Importing Data Files
path = r'/Users/aya/Desktop/DataAnalytics/Data-Immersion/Achievement-4/Instacart-Basket-Analysis'

In [5]:
path

'/Users/aya/Desktop/DataAnalytics/Data-Immersion/Achievement-4/Instacart-Basket-Analysis'

# 03. Importing Data

In [7]:
#Import ords_prods_merge_Deriving_Variables Dataset
ords_prods_merge = pd.read_pickle(os.path.join(path, '02-Data', 'Prepared-Data', 'ords_prods_merge_Deriving_Variables.pkl'))

In [8]:
#Checking the dataframe
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_since_prior_order  add_to_cart_order  reordered _merge  \
0                     3.0                  5          0   both   
1                    20.0                  1          1   both   
2                     6.0                 20          0   both   
3                     NaN                 10          0   both   
4                     9.0                 11          1   both   

     price_range_loc     busiest_day  Busiest_days busiest_period_of_day  
0  Mid-range product  Regularly busy  Regular days           Most orders  
1  Mid-range product  Regularly busy  Regular days        Average orders  
2  Mid-range product     Busiest day  Busiest days        Average orders  
3  Mid-range product  Regularly busy  Slowest days           Most orders  
4  Mid-range product      Least busy  Slowest days        Average orders

# 04. Data Manipulation

In [10]:
#Subsetting the 'ords_prods_merge' dataframe and assign it to 'df' dataframe
df = ords_prods_merge[:1000000]

In [11]:
#Checking the the 'df' dataframe dimension
df.shape

(1000000, 18)

In [12]:
#Checking the the 'df' dataframe
df.head(10)

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   
5           1  Chocolate Sandwich Cookies        61             19     5.8   
6           1  Chocolate Sandwich Cookies        61             19     5.8   
7           1  Chocolate Sandwich Cookies        61             19     5.8   
8           1  Chocolate Sandwich Cookies        61             19     5.8   
9           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   
5   1701441      777            16                   1                  7   
6   1871483      825             3                   2                 14   
7   1290456      910            12                   3                 10   
8    369558     1052            10                   1                 20   
9    589712     1052            15                   1                 12   

   days_since_prior_order  add_to_cart_order  reordered _merge  \
0                     3.0                  5          0   both   
1                    20.0                  1          1   both   
2                     6.0                 20          0   both   
3                     NaN                 10          0   both   
4                     9.0                 11          1   both   
5                    26.0                  7          0   both   
6                    30.0                  2          0   both   
7                    30.0                  1          0   both   
8                    19.0                  1          0   both   
9                    15.0                  2          1   both   

     price_range_loc     busiest_day  Busiest_days busiest_period_of_day  
0  Mid-range product  Regularly busy  Regular days           Most orders  
1  Mid-range product  Regularly busy  Regular days        Average orders  
2  Mid-range product     Busiest day  Busiest days        Average orders  
3  Mid-range product  Regularly busy  Slowest days           Most orders  
4  Mid-range product      Least busy  Slowest days        Average orders  
5  Mid-range product  Regularly busy  Busiest days        Average orders  
6  Mid-range product  Regularly busy  Regular days           Most orders  
7  Mid-range product  Regularly busy  Slowest days           Most orders  
8  Mid-range product  Regularly busy  Busiest days        Average orders  
9  Mid-range product  Regularly busy  Busiest days           Most orders

# 05. Grouping Data with pandas

In [14]:
#Grouping the 'df' dataframe by 'product_name'
df.groupby('product_name')

# 06. Aggregating Data

In [16]:
#Single Aggregation (mean)
#Aggregate the 'order_number' column in the 'df' dataframe grouped by 'department_id'
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.577493
2                17.320781
3                16.084944
4                17.530458
5                14.763075
6                16.658449
7                17.031590
8                15.076662
9                15.447580
10               18.681852
11               15.447411
12               14.327957
13               16.548642
14               16.960241
15               16.121948
16               17.803851
17               15.593633
18               19.674252
19               16.899756
20               16.255442
21               25.535479

In [17]:
#Multiple Aggregations (mean', 'min', 'max)
#Aggregate the 'order_number' column in the 'df' dataframe grouped by 'department_id' 
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
1                15.577493   1  99
2                17.320781   1  96
3                16.084944   1  99
4                17.530458   1  99
5                14.763075   1  99
6                16.658449   1  99
7                17.031590   1  99
8                15.076662   1  98
9                15.447580   1  99
10               18.681852   1  99
11               15.447411   1  99
12               14.327957   1  99
13               16.548642   1  99
14               16.960241   1  99
15               16.121948   1  99
16               17.803851   1  99
17               15.593633   1  99
18               19.674252   1  99
19               16.899756   1  99
20               16.255442   1  99
21               25.535479   1  98

# 07. Aggregating Data with transform()

In [19]:
#Creating a new column named 'max_order' in the 'df' dataframe, Grouped by 'user_id' column &
#Applying the transform function on the 'order_number' column to Aggregate by Max. 
df['max_order'] = df.groupby(['user_id'])['order_number'].transform(np.max)

/var/folders/62/xltcnc5n7d97j6t5ql_87vym0000gn/T/ipykernel_14577/3541529635.py:3: FutureWarning: The provided callable <function max at 0x1060f3060> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  df['max_order'] = df.groupby(['user_id'])['order_number'].transform(np.max)
/var/folders/62/xltcnc5n7d97j6t5ql_87vym0000gn/T/ipykernel_14577/3541529635.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['max_order'] = df.groupby(['user_id'])['order_number'].transform(np.max)


In [20]:
#Checking the 'df' dataframe
df.head(10)

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   
5           1  Chocolate Sandwich Cookies        61             19     5.8   
6           1  Chocolate Sandwich Cookies        61             19     5.8   
7           1  Chocolate Sandwich Cookies        61             19     5.8   
8           1  Chocolate Sandwich Cookies        61             19     5.8   
9           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   
5   1701441      777            16                   1                  7   
6   1871483      825             3                   2                 14   
7   1290456      910            12                   3                 10   
8    369558     1052            10                   1                 20   
9    589712     1052            15                   1                 12   

   days_since_prior_order  add_to_cart_order  reordered _merge  \
0                     3.0                  5          0   both   
1                    20.0                  1          1   both   
2                     6.0                 20          0   both   
3                     NaN                 10          0   both   
4                     9.0                 11          1   both   
5                    26.0                  7          0   both   
6                    30.0                  2          0   both   
7                    30.0                  1          0   both   
8                    19.0                  1          0   both   
9                    15.0                  2          1   both   

     price_range_loc     busiest_day  Busiest_days busiest_period_of_day  \
0  Mid-range product  Regularly busy  Regular days           Most orders   
1  Mid-range product  Regularly busy  Regular days        Average orders   
2  Mid-range product     Busiest day  Busiest days        Average orders   
3  Mid-range product  Regularly busy  Slowest days           Most orders   
4  Mid-range product      Least busy  Slowest days        Average orders   
5  Mid-range product  Regularly busy  Busiest days        Average orders   
6  Mid-range product  Regularly busy  Regular days           Most orders   
7  Mid-range product  Regularly busy  Slowest days           Most orders   
8  Mid-range product  Regularly busy  Busiest days        Average orders   
9  Mid-range product  Regularly busy  Busiest days           Most orders   

   max_order  
0         30  
1         30  
2          5  
3          3  
4          3  
5         16  
6          3  
7         12  
8         18  
9         18

In [21]:
#Sorting the dataframe by 'user_id' to confirm the number of orders with the maximum order
df.sort_values(by=['user_id'], ascending=[True]).head(100)

product_id                                       product_name  \
92369          196                                               Soda   
92375          196                                               Soda   
92370          196                                               Soda   
92374          196                                               Soda   
92373          196                                               Soda   
...            ...                                                ...   
494265        1005                         China Green Tips Green Tea   
494264        1005                         China Green Tips Green Tea   
494263        1005                         China Green Tips Green Tea   
656370        1323                 Vanilla Bean Natural Lowfat Yogurt   
985637        2067  Plus Cranberry Almond + Antioxidants with Maca...   

        aisle_id  department_id  prices  order_id  user_id  order_number  \
92369         77              7     9.0   2398795        1             2   
92375         77              7     9.0   3108588        1             8   
92370         77              7     9.0    473747        1             3   
92374         77              7     9.0    550135        1             7   
92373         77              7     9.0   3367565        1             6   
...          ...            ...     ...       ...      ...           ...   
494265        94              7     4.7   1149132       27            21   
494264        94              7     4.7   2705360       27            16   
494263        94              7     4.7   1406911       27            14   
656370       120             16    11.1   1603399       27            38   
985637         3             19     5.5   2005447       27            37   

        orders_day_of_week  order_hour_of_day  days_since_prior_order  \
92369                    3                  7                    15.0   
92375                    1                 14                    14.0   
92370                    3                 12                    21.0   
92374                    1                  9                    20.0   
92373                    2                  7                    19.0   
...                    ...                ...                     ...   
494265                   2                 11                     8.0   
494264                   2                 16                     8.0   
494263                   3                  9                     8.0   
656370                   2                 16                     1.0   
985637                   1                 10                     3.0   

        add_to_cart_order  reordered _merge    price_range_loc  \
92369                   1          1   both  Mid-range product   
92375                   2          1   both  Mid-range product   
92370                   1          1   both  Mid-range product   
92374                   1          1   both  Mid-range product   
92373                   1          1   both  Mid-range product   
...                   ...        ...    ...                ...   
494265                  7          1   both  Low-range product   
494264                  9          1   both  Low-range product   
494263                  8          0   both  Low-range product   
656370                  8          0   both  Mid-range product   
985637                  4          0   both  Mid-range product   

           busiest_day  Busiest_days busiest_period_of_day  max_order  
92369   Regularly busy  Slowest days        Average orders         10  
92375   Regularly busy  Busiest days           Most orders         10  
92370   Regularly busy  Slowest days           Most orders         10  
92374   Regularly busy  Busiest days           Most orders         10  
92373   Regularly busy  Regular days        Average orders         10  
...                ...           ...                   ...        ...  
494265  Regularly busy  Regular days        

In [22]:
#Chckeing the 'df' dataframe
df.head(100)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
..         ...                         ...       ...            ...     ...   
95           1  Chocolate Sandwich Cookies        61             19     5.8   
96           1  Chocolate Sandwich Cookies        61             19     5.8   
97           1  Chocolate Sandwich Cookies        61             19     5.8   
98           1  Chocolate Sandwich Cookies        61             19     5.8   
99           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    3139998      138            28                   6                 11   
1    1977647      138            30                   6                 17   
2     389851      709             2                   0                 21   
3     652770      764             1                   3                 13   
4    1813452      764             3                   4                 17   
..       ...      ...           ...                 ...                ...   
95    602103    10831             8                   3                 11   
96     49629    11119            18                   1                 14   
97    317888    11186            13                   5                 16   
98    682486    11243            16                   3                 13   
99   2861386    11448            10                   2                 15   

    days_since_prior_order  add_to_cart_order  reordered _merge  \
0                      3.0                  5          0   both   
1                     20.0                  1          1   both   
2                      6.0                 20          0   both   
3                      NaN                 10          0   both   
4                      9.0                 11          1   both   
..                     ...                ...        ...    ...   
95                    23.0                  5          0   both   
96                    30.0                  1          0   both   
97                     2.0                  8          0   both   
98                     0.0                  2          0   both   
99                     3.0                  6          0   both   

      price_range_loc     busiest_day  Busiest_days busiest_period_of_day  \
0   Mid-range product  Regularly busy  Regular days           Most orders   
1   Mid-range product  Regularly busy  Regular days        Average orders   
2   Mid-range product     Busiest day  Busiest days        Average orders   
3   Mid-range product  Regularly busy  Slowest days           Most orders   
4   Mid-range product      Least busy  Slowest days        Average orders   
..                ...             ...           ...                   ...   
95  Mid-range product  Regularly busy  Slowest days           Most orders   
96  Mid-range product  Regularly busy  Busiest days           Most orders   
97  Mid-range product  Regularly busy  Regular days           Most orders   
98  Mid-range product  Regularly busy  Slowest days           Most orders   
99  Mid-range product  Regularly busy  Regular days           Most orders   

    max_order  
0          30  
1          30  
2           5  
3           3  
4           3  
..        ...  
95          8  
96         20  
97         23  
98         16  
99         21  

[100 rows x 19 columns]

In [23]:
#Changing the output display - to be able to view the desired bmuber of rows. 
pd.options.display.max_rows = None

In [24]:
#Chckeing the 'df' dataframe Again
df.head(100)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   
15           1  Chocolate Sandwich Cookies        61             19     5.8   
16           1  Chocolate Sandwich Cookies        61             19     5.8   
17           1  Chocolate Sandwich Cookies        61             19     5.8   
18           1  Chocolate Sandwich Cookies        61             19     5.8   
19           1  Chocolate Sandwich Cookies        61             19     5.8   
20           1  Chocolate Sandwich Cookies        61             19     5.8   
21           1  Chocolate Sandwich Cookies        61             19     5.8   
22           1  Chocolate Sandwich Cookies        61             19     5.8   
23           1  Chocolate Sandwich Cookies        61             19     5.8   
24           1  Chocolate Sandwich Cookies        61             19     5.8   
25           1  Chocolate Sandwich Cookies        61             19     5.8   
26           1  Chocolate Sandwich Cookies        61             19     5.8   
27           1  Chocolate Sandwich Cookies        61             19     5.8   
28           1  Chocolate Sandwich Cookies        61             19     5.8   
29           1  Chocolate Sandwich Cookies        61             19     5.8   
30           1  Chocolate Sandwich Cookies        61             19     5.8   
31           1  Chocolate Sandwich Cookies        61             19     5.8   
32           1  Chocolate Sandwich Cookies        61             19     5.8   
33           1  Chocolate Sandwich Cookies        61             19     5.8   
34           1  Chocolate Sandwich Cookies        61             19     5.8   
35           1  Chocolate Sandwich Cookies        61             19     5.8   
36           1  Chocolate Sandwich Cookies        61             19     5.8   
37           1  Chocolate Sandwich Cookies        61             19     5.8   
38           1  Chocolate Sandwich Cookies        61             19     5.8   
39           1  Chocolate Sandwich Cookies        61             19     5.8   
40           1  Chocolate Sandwich Cookies        61             19     5.8   
41           1  Chocolate Sandwich Cookies        61             19     5.8   
42           1  Chocolate Sandwich Cookies        61             19     5.8   
43           1  Chocolate Sandwich Cookies        61             19     5.8   
44           1  Chocolate Sandwich Cookies        61             19     5.8   
45           1  Chocolate Sandwich Cookies        61             19     5.8   
46           1  Chocolate Sandwich Cookies        61             19     5.8   
47           1  Chocolate Sandwich Cookies        61             19     5.8   
48           1  Chocolate Sandwich Cookies        61             19     5.8   
49           1  Chocolate Sandwich Cookies        61  

# 08. Deriving Columns with loc()
##### Creating falgs - Assinging loyalty status in the new column named 'loyalty_flag' based on the rows of the 'max_order' column in the 'ords_prods_merge' dataframe

In [26]:
#Assinging loyalty status - (> 40)  then 'Loyal customer'
df.loc[df['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

/var/folders/62/xltcnc5n7d97j6t5ql_87vym0000gn/T/ipykernel_14577/3336362628.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'


In [27]:
#Assinging loyalty status -  (<= 40 &  > 10) then 'Regular customer'
df.loc[(df['max_order'] <= 40) & (df['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [28]:
#Assinging loyalty status - (<= 10) then 'New customer'
df.loc[df['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [29]:
#Checking the new column 
df['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    473070
Loyal customer      287917
New customer        239013
Name: count, dtype: int64

In [30]:
#Checking the a list of columns within a the dataframe
df[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0       138  Regular customer            28
1       138  Regular customer            30
2       709      New customer             2
3       764      New customer             1
4       764      New customer             3
5       777  Regular customer            16
6       825      New customer             3
7       910  Regular customer            12
8      1052  Regular customer            10
9      1052  Regular customer            15
10     1379      New customer             1
11     1480    Loyal customer            36
12     1480    Loyal customer            38
13     1480    Loyal customer            39
14     1494      New customer             6
15     1494      New customer             7
16     1494      New customer             9
17     1540  Regular customer             4
18     1540  Regular customer             5
19     1540  Regular customer             6
20     1540  Regular customer             7
21     1540  Regular customer             8
22     1540  Regular customer             9
23     1540  Regular customer            14
24     1540  Regular customer            15
25     1540  Regular customer            18
26     1540  Regular customer            19
27     1540  Regular customer            20
28     1540  Regular customer            21
29     1540  Regular customer            22
30     1540  Regular customer            24
31     1540  Regular customer            25
32     1540  Regular customer            27
33     1540  Regular customer            28
34     1598      New customer             1
35     1598      New customer             2
36     1598      New customer             3
37     1647      New customer             1
38     1647      New customer             4
39     1647      New customer             6
40     1647      New customer             8
41     1647      New customer             9
42     2833      New customer             1
43     2850  Regular customer             7
44     2857  Regular customer            24
45     3010      New customer             5
46     3029      New customer             1
47     3395      New customer             4
48     3904      New customer             1
49     4006  Regular customer             2
50     4006  Regular customer             3
51     4006  Regular customer             5
52     4006  Regular customer             9
53     4006  Regular customer            12
54     4122  Regular customer            11
55     4141      New customer             2
56     4343      New customer             1
57     4553      New customer             3
58     4845      New customer             6
59     6866      New customer             1

# Task 4.8

In [32]:
#Checking the ords_prods_merge dataframe
ords_prods_merge.shape

(32404859, 18)

In [33]:
##Aggregate the 'order_number' column in the ords_prods_merge dataframe grouped by 'department_id'
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

# Observation 2. & 3.
###### Average orders for each department (There are 21 departments)
###### The range of means between different departments is between ~15 to ~22. 

In [35]:
#Creating a new column named 'max_order' in the 'ords_prods_merge' dataframe, Grouped by 'user_id' column &
#Applying the transform function on the 'order_number' column to Aggregate by Max. 
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

/var/folders/62/xltcnc5n7d97j6t5ql_87vym0000gn/T/ipykernel_14577/2351130679.py:3: FutureWarning: The provided callable <function max at 0x1060f3060> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [36]:
#Checking the dataframe
ords_prods_merge.shape

(32404859, 19)

In [37]:
#Checking the dataframe
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_since_prior_order  add_to_cart_order  reordered _merge  \
0                     3.0                  5          0   both   
1                    20.0                  1          1   both   
2                     6.0                 20          0   both   
3                     NaN                 10          0   both   
4                     9.0                 11          1   both   

     price_range_loc     busiest_day  Busiest_days busiest_period_of_day  \
0  Mid-range product  Regularly busy  Regular days           Most orders   
1  Mid-range product  Regularly busy  Regular days        Average orders   
2  Mid-range product     Busiest day  Busiest days        Average orders   
3  Mid-range product  Regularly busy  Slowest days           Most orders   
4  Mid-range product      Least busy  Slowest days        Average orders   

   max_order  
0         32  
1         32  
2          5  
3          3  
4          3

In [38]:
#Deriving Columns
#Assinging loyalty status - (> 40)  then 'Loyal customer'
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [39]:
#Assinging loyalty status -  (<= 40 &  > 10) then 'Regular customer'
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [40]:
#Assinging loyalty status - (<= 10) then 'New customer'
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [41]:
#Checking the new column 
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [42]:
#Checking the a list of columns within a the dataframe
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0       138  Regular customer            28
1       138  Regular customer            30
2       709      New customer             2
3       764      New customer             1
4       764      New customer             3
5       777  Regular customer            16
6       825      New customer             3
7       910  Regular customer            12
8      1052  Regular customer            10
9      1052  Regular customer            15
10     1379      New customer             1
11     1480    Loyal customer            36
12     1480    Loyal customer            38
13     1480    Loyal customer            39
14     1494  Regular customer             6
15     1494  Regular customer             7
16     1494  Regular customer             9
17     1540  Regular customer             4
18     1540  Regular customer             5
19     1540  Regular customer             6
20     1540  Regular customer             7
21     1540  Regular customer             8
22     1540  Regular customer             9
23     1540  Regular customer            14
24     1540  Regular customer            15
25     1540  Regular customer            18
26     1540  Regular customer            19
27     1540  Regular customer            20
28     1540  Regular customer            21
29     1540  Regular customer            22
30     1540  Regular customer            24
31     1540  Regular customer            25
32     1540  Regular customer            27
33     1540  Regular customer            28
34     1598      New customer             1
35     1598      New customer             2
36     1598      New customer             3
37     1647      New customer             1
38     1647      New customer             4
39     1647      New customer             6
40     1647      New customer             8
41     1647      New customer             9
42     2833      New customer             1
43     2850  Regular customer             7
44     2857  Regular customer            24
45     3010  Regular customer             5
46     3029      New customer             1
47     3395      New customer             4
48     3904      New customer             1
49     4006    Loyal customer             2
50     4006    Loyal customer             3
51     4006    Loyal customer             5
52     4006    Loyal customer             9
53     4006    Loyal customer            12
54     4122  Regular customer            11
55     4141      New customer             2
56     4343      New customer             1
57     4553      New customer             3
58     4845  Regular customer             6
59     6866      New customer             1

# Observation 4.
#### The entrier dataframe ords_prods_merge:
#### Groued by 'user_id'
#### Aggregated the 'order_number' and transform(max)
#### Assigned flags using the loc()

In [44]:
#Checking the dataframe 'ords_prods_merge'
ords_prods_merge.head(20)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   
15           1  Chocolate Sandwich Cookies        61             19     5.8   
16           1  Chocolate Sandwich Cookies        61             19     5.8   
17           1  Chocolate Sandwich Cookies        61             19     5.8   
18           1  Chocolate Sandwich Cookies        61             19     5.8   
19           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    3139998      138            28                   6                 11   
1    1977647      138            30                   6                 17   
2     389851      709             2                   0                 21   
3     652770      764             1                   3                 13   
4    1813452      764             3                   4                 17   
5    1701441      777            16                   1                  7   
6    1871483      825             3                   2                 14   
7    1290456      910            12                   3                 10   
8     369558     1052            10                   1                 20   
9     589712     1052            15                   1                 12   
10   1798115     1379             1                   5                 11   
11    497193     1480            36                   3                 13   
12   1727826     1480            38                   4                 13   
13   2694698     1480            39                   2                 12   
14   3171753     1494             6                   3                 10   
15    186761     1494             7                   2                  8   
16    317121     1494             9                   5                  8   
17    200390     1540             4                   2                  0   
18    374436     1540             5                   1                 21   
19   2920553     1540             6                   1                  0   

    days_since_prior_order  add_to_cart_order  reordered _merge  \
0                      3.0                  5          0   both   
1                     20.0                  1          1   both   
2                      6.0                 20          0   both   
3                      NaN                 10          0   both   
4                      9.0                 11          1   both   
5                     26.0                  7          0   both   
6                     30.0                  2          0   both   
7                     30.0                  1          0   both   
8                     19.0                  1          0   both   
9                     15.0         

In [45]:
#Aggregate the 'prices' column in the 'ords_prods_merge' dataframe grouped by 'loyalty_flag' 
#The basic statsitcs of the product prices of each loyalty category
ords_prods_merge.groupby(['loyalty_flag']).agg({'prices':['mean', 'min', 'max']})

prices           
                      mean  min   max
loyalty_flag                         
Loyal customer    7.773487  1.0  25.0
New customer      7.801077  1.0  25.0
Regular customer  7.798145  1.0  25.0

# Observation 5. 
##### The prices of products purchased by loyal customers are  comparable or very close to those purchased by regular or new customers.

In [47]:
#Apply the transform () funtion on the 'prices' column to generate the average spent amount for each user
#Assign the results in a new column names 'Avg_spending'
ords_prods_merge['Avg_spending'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

/var/folders/62/xltcnc5n7d97j6t5ql_87vym0000gn/T/ipykernel_14577/4152911572.py:3: FutureWarning: The provided callable <function mean at 0x1060f3a60> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  ords_prods_merge['Avg_spending'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [48]:
#Checkign the dataframe
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   ...  add_to_cart_order  reordered  _merge    price_range_loc  \
0  ...                  5          0    both  Mid-range product   
1  ...                  1          1    both  Mid-range product   
2  ...                 20          0    both  Mid-range product   
3  ...                 10          0    both  Mid-range product   
4  ...                 11          1    both  Mid-range product   

      busiest_day  Busiest_days busiest_period_of_day max_order  \
0  Regularly busy  Regular days           Most orders        32   
1  Regularly busy  Regular days        Average orders        32   
2     Busiest day  Busiest days        Average orders         5   
3  Regularly busy  Slowest days           Most orders         3   
4      Least busy  Slowest days        Average orders         3   

       loyalty_flag Avg_spending  
0  Regular customer     6.935811  
1  Regular customer     6.935811  
2      New customer     7.930208  
3      New customer     4.972414  
4      New customer     4.972414  

[5 rows x 21 columns]

In [49]:
#Deriving anew column named 'spending_flag', with loc()
#Setting up criteria - High spender >= 10
ords_prods_merge.loc[ords_prods_merge['Avg_spending'] >= 10, 'spending_flag'] = 'High spender'

In [50]:
#Deriving anew column named 'spending_flag', with loc()
#Setting up criteria - Low spender < 10
ords_prods_merge.loc[ords_prods_merge['Avg_spending'] < 10, 'spending_flag'] = 'Low spender'

In [51]:
#Checking the frequency of the new column in the dataframe 'ords_prods_merge'
ords_prods_merge['spending_flag'].value_counts(dropna = False)

spending_flag
Low spender     32285150
High spender      119709
Name: count, dtype: int64

In [52]:
#Checking the dimention of the dataframe 'ords_prods_merge'
ords_prods_merge.shape

(32404859, 22)

In [53]:
#Checking the dataframe based on the new coumn
ords_prods_merge[['user_id', 'loyalty_flag', 'spending_flag', 'Avg_spending']].head(60)

user_id      loyalty_flag spending_flag  Avg_spending
0       138  Regular customer   Low spender      6.935811
1       138  Regular customer   Low spender      6.935811
2       709      New customer   Low spender      7.930208
3       764      New customer   Low spender      4.972414
4       764      New customer   Low spender      4.972414
5       777  Regular customer   Low spender      6.935398
6       825      New customer   Low spender      5.957576
7       910  Regular customer   Low spender      6.680000
8      1052  Regular customer   Low spender      7.162500
9      1052  Regular customer   Low spender      7.162500
10     1379      New customer   Low spender      5.813636
11     1480    Loyal customer   Low spender      5.081633
12     1480    Loyal customer   Low spender      5.081633
13     1480    Loyal customer   Low spender      5.081633
14     1494  Regular customer   Low spender      6.803797
15     1494  Regular customer   Low spender      6.803797
16     1494  Regular customer   Low spender      6.803797
17     1540  Regular customer   Low spender      6.607122
18     1540  Regular customer   Low spender      6.607122
19     1540  Regular customer   Low spender      6.607122
20     1540  Regular customer   Low spender      6.607122
21     1540  Regular customer   Low spender      6.607122
22     1540  Regular customer   Low spender      6.607122
23     1540  Regular customer   Low spender      6.607122
24     1540  Regular customer   Low spender      6.607122
25     1540  Regular customer   Low spender      6.607122
26     1540  Regular customer   Low spender      6.607122
27     1540  Regular customer   Low spender      6.607122
28     1540  Regular customer   Low spender      6.607122
29     1540  Regular customer   Low spender      6.607122
30     1540  Regular customer   Low spender      6.607122
31     1540  Regular customer   Low spender      6.607122
32     1540  Regular customer   Low spender      6.607122
33     1540  Regular customer   Low spender      6.607122
34     1598      New customer   Low spender      7.123529
35     1598      New customer   Low spender      7.123529
36     1598      New customer   Low spender      7.123529
37     1647      New customer   Low spender      4.853521
38     1647      New customer   Low spender      4.853521
39     1647      New customer   Low spender      4.853521
40     1647      New customer   Low spender      4.853521
41     1647      New customer   Low spender      4.853521
42     2833      New customer   Low spender      7.377143
43     2850  Regular customer   Low spender      8.054153
44     2857  Regular customer   Low spender      8.353793
45     3010  Regular customer   Low spender      6.452800
46     3029      New customer   Low spender      7.960606
47     3395      New customer   Low spender      6.322222
48     3904      New customer   Low spender      7.511111
49     4006    Loyal customer   Low spender      7.441100
50     4006    Loyal customer   Low spender      7.441100
51     4006    Loyal customer   Low spender      7.441100
52     4006    Loyal customer   Low spender      7.441100
53     4006    Loyal customer   Low spender      7.441100
54     4122  Regular customer   Low spender      7.598765
55     4141      New customer   Low spender      6.033333
56     4343      New customer   Low spender      7.393939
57     4553      New customer   Low spender      5.715789
58     4845  Regular customer   Low spender      7.229630
59     6866      New customer   Low spender      4.807143

In [54]:
#Checking the frequency of spending habits based on loyalty status.
ords_prods_merge.groupby('loyalty_flag')['spending_flag'].value_counts()

loyalty_flag      spending_flag
Loyal customer    Low spender      10270197
                  High spender        13896
New customer      Low spender       6185362
                  High spender        58628
Regular customer  Low spender      15829591
                  High spender        47185
Name: count, dtype: int64

# Observation 6. 

##### Checking the frequency of the spending
##### Created a new column named 'Avg_spending'
##### Splitting the data into groups based on the 'user_id' column
##### Applying the transform() function on the 'prices' column to calculate the average amount spent by each user
##### Creating a new column named 'Avg_spending' to store the results of the aggregation


##### Created 'spending_flag' based on the average amount spent per customer. 
##### If a customer spends more than or equal to 10, they are classified as a high spender. 
##### If a customer spends less than 10, they are classified as a low spender."


##### We can see that there are more'low spender' customers than 'high spender' customers. 
##### When considering the frequency of spending habits based on loyalty status: the "Loyal customers" do not necessarily drive sales in Instacart. 
##### Highest spenders: 'new customers', followed by 'regular customers', and then 'loyal customers'. 
##### Lowest spenders: 'regular spenders', followed by 'loyal customers', and then 'new customers'.

##### Total dataframe ords_prods_merge 32404859 rows X 22 columns

In [56]:
pd.options.display.max_columns = None

In [57]:
#Checking the dataframe 'ords_prods_merge'
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_since_prior_order  add_to_cart_order  reordered _merge  \
0                     3.0                  5          0   both   
1                    20.0                  1          1   both   
2                     6.0                 20          0   both   
3                     NaN                 10          0   both   
4                     9.0                 11          1   both   

     price_range_loc     busiest_day  Busiest_days busiest_period_of_day  \
0  Mid-range product  Regularly busy  Regular days           Most orders   
1  Mid-range product  Regularly busy  Regular days        Average orders   
2  Mid-range product     Busiest day  Busiest days        Average orders   
3  Mid-range product  Regularly busy  Slowest days           Most orders   
4  Mid-range product      Least busy  Slowest days        Average orders   

   max_order      loyalty_flag  Avg_spending spending_flag  
0         32  Regular customer      6.935811   Low spender  
1         32  Regular customer      6.935811   Low spender  
2          5      New customer      7.930208   Low spender  
3          3      New customer      4.972414   Low spender  
4          3      New customer      4.972414   Low spender

In [58]:
#Apply the transform () funtion on the 'days_since_prior_order' column to generate the median 'days_since_prior_order' for each user
#Assign the results in a new column names 'median_days_since_last_order'
ords_prods_merge['median_days_since_last_order'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

/var/folders/62/xltcnc5n7d97j6t5ql_87vym0000gn/T/ipykernel_14577/2339524792.py:3: FutureWarning: The provided callable <function median at 0x10621f100> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  ords_prods_merge['median_days_since_last_order'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [59]:
#Checking the dataframe 'ords_prods_merge'
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_since_prior_order  add_to_cart_order  reordered _merge  \
0                     3.0                  5          0   both   
1                    20.0                  1          1   both   
2                     6.0                 20          0   both   
3                     NaN                 10          0   both   
4                     9.0                 11          1   both   

     price_range_loc     busiest_day  Busiest_days busiest_period_of_day  \
0  Mid-range product  Regularly busy  Regular days           Most orders   
1  Mid-range product  Regularly busy  Regular days        Average orders   
2  Mid-range product     Busiest day  Busiest days        Average orders   
3  Mid-range product  Regularly busy  Slowest days           Most orders   
4  Mid-range product      Least busy  Slowest days        Average orders   

   max_order      loyalty_flag  Avg_spending spending_flag  \
0         32  Regular customer      6.935811   Low spender   
1         32  Regular customer      6.935811   Low spender   
2          5      New customer      7.930208   Low spender   
3          3      New customer      4.972414   Low spender   
4          3      New customer      4.972414   Low spender   

   median_days_since_last_order  
0                           8.0  
1                           8.0  
2                           8.0  
3                           9.0  
4                           9.0

In [60]:
#Deriving anew column named 'order_frquency_flag', with loc()
#Setting up criteria - 'Non-frequent customer' > 20
ords_prods_merge.loc[ords_prods_merge['median_days_since_last_order'] > 20, 'order_frquency_flag'] = 'Non-frequent customer'

In [61]:
#Deriving anew column named 'order_frquency_flag', with loc()
#Setting up criteria - 'Regular customer' <= 20 &  > 10
ords_prods_merge.loc[(ords_prods_merge['median_days_since_last_order'] <= 20) & (ords_prods_merge['median_days_since_last_order'] > 10), 'order_frquency_flag'] = 'Regular customer'

In [62]:
#Deriving anew column named 'order_frquency_flag', with loc()
#Setting up criteria - 'Frequent customer' <= 10
ords_prods_merge.loc[ords_prods_merge['median_days_since_last_order'] <= 10, 'order_frquency_flag'] = 'Frequent customer'

In [63]:
#Checking the frequency of the new column in the dataframe 'ords_prods_merge'
ords_prods_merge['order_frquency_flag'].value_counts(dropna = False)

order_frquency_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: count, dtype: int64

In [64]:
#Checking the dimention of the dataframe 'ords_prods_merge'
ords_prods_merge.shape

(32404859, 24)

In [65]:
#Checking the dataframe based on the new coumn
ords_prods_merge[['user_id', 'loyalty_flag', 'spending_flag', 'order_frquency_flag', 'median_days_since_last_order']].head(60)

user_id      loyalty_flag spending_flag    order_frquency_flag  \
0       138  Regular customer   Low spender      Frequent customer   
1       138  Regular customer   Low spender      Frequent customer   
2       709      New customer   Low spender      Frequent customer   
3       764      New customer   Low spender      Frequent customer   
4       764      New customer   Low spender      Frequent customer   
5       777  Regular customer   Low spender       Regular customer   
6       825      New customer   Low spender       Regular customer   
7       910  Regular customer   Low spender      Frequent customer   
8      1052  Regular customer   Low spender      Frequent customer   
9      1052  Regular customer   Low spender      Frequent customer   
10     1379      New customer   Low spender       Regular customer   
11     1480    Loyal customer   Low spender      Frequent customer   
12     1480    Loyal customer   Low spender      Frequent customer   
13     1480    Loyal customer   Low spender      Frequent customer   
14     1494  Regular customer   Low spender      Frequent customer   
15     1494  Regular customer   Low spender      Frequent customer   
16     1494  Regular customer   Low spender      Frequent customer   
17     1540  Regular customer   Low spender      Frequent customer   
18     1540  Regular customer   Low spender      Frequent customer   
19     1540  Regular customer   Low spender      Frequent customer   
20     1540  Regular customer   Low spender      Frequent customer   
21     1540  Regular customer   Low spender      Frequent customer   
22     1540  Regular customer   Low spender      Frequent customer   
23     1540  Regular customer   Low spender      Frequent customer   
24     1540  Regular customer   Low spender      Frequent customer   
25     1540  Regular customer   Low spender      Frequent customer   
26     1540  Regular customer   Low spender      Frequent customer   
27     1540  Regular customer   Low spender      Frequent customer   
28     1540  Regular customer   Low spender      Frequent customer   
29     1540  Regular customer   Low spender      Frequent customer   
30     1540  Regular customer   Low spender      Frequent customer   
31     1540  Regular customer   Low spender      Frequent customer   
32     1540  Regular customer   Low spender      Frequent customer   
33     1540  Regular customer   Low spender      Frequent customer   
34     1598      New customer   Low spender      Frequent customer   
35     1598      New customer   Low spender      Frequent customer   
36     1598      New customer   Low spender      Frequent customer   
37     1647      New customer   Low spender  Non-frequent customer   
38     1647      New customer   Low spender  Non-frequent customer   
39     1647      New customer   Low spender  Non-frequent customer   
40     1647      New customer   Low spender  Non-frequent customer   
41     1647      New customer   Low spender  Non-frequent customer   
42     2833      New customer   Low spender       Regular customer   
43     2850  Regular customer   Low spender      Frequent customer   
44     2857  Regular customer   Low spender       Regular customer   
45     3010  Regular customer   Low spender      Frequent customer   
46     3029      New customer   Low spender  Non-frequent customer   
47     3395      New customer   Low spender       Regular customer   
48     3904      New customer   Low spender  Non-frequent customer   
49     4006    Loyal customer   Low spender      Frequent customer   
50     4006    Loyal customer   Low spender      Frequent customer   
51     4006    Loyal customer   Low spender      Frequent customer   
52     4006    Loyal customer   Low spender      Frequent customer   
53     4006    Loyal customer   Low spender      Frequent customer   
54     4122  Regular customer   Low spender      Frequent customer   
55     4141      New customer   Low spender  Non-frequent customer   
56     4343   

In [66]:
#Checking the order frequency based on spending status.
ords_prods_merge.groupby('spending_flag')['order_frquency_flag'].value_counts()

spending_flag  order_frquency_flag  
High spender   Frequent customer           56944
               Regular customer            33234
               Non-frequent customer       29531
Low spender    Frequent customer        21502909
               Regular customer          7175330
               Non-frequent customer     3606906
Name: count, dtype: int64

# Observation 7. 

##### Checking the frequency of the spending


##### Checking the frequency of days since placing the last order.
##### Created a new column named 'median_days_since_last_order'
##### Splitting the data into groups based on the 'user_id' column
##### Applying the transform() function on the 'days_since_prior_order' column to calculate the median days since the last order by each user
##### Creating a new column named 'median_days_since_last_order' to store the results of the aggregation


##### Created 'order_frquency_flag' based on the 'median_days_since_last_order' per customer. 
##### If a customer 'median_days_since_last_order',  > 20, they are classified as 'Non-frequent customer'. 
##### If a customer 'median_days_since_last_order', <= 20 & > 10, they are classified as 'Regular customer'. 
##### If a customer 'median_days_since_last_order', <= 10, they are classified as 'Frequent customer'. 

##### We can see that there are more'Frequent customer' customers than other customers. 
##### The results also show 'NaN' where there is no available data for the very first order placed by the customers.
##### When considering the order frequency based on spending status: the "High spender" do not necessarily placing frequent orders in Instacart. 
##### Lower spender customers are 'Frequent customer' compared to High spender customers.

##### Total dataframe ords_prods_merge 32404859 rows X 24 columns.

In [68]:
# Export the 'ords_prods_merge_agg' dataframe to pkl formate
ords_prods_merge.to_pickle(os.path.join(path, '02-Data','Prepared-Data', 'ords_prods_merge_agg.pkl'))

### Observation 9.
##### Export the grouped and aggregated data in pickle format as “ords_prods_merge_agg.pkl” 